File: backend/app/schemas/report.py
"""报告域响应契约：建议报告与回测任务。"""

from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

__all__ = [
    "AdviceLatestResponse",
    "AdviceMarkOut",
    "AdviceMarkRequest",
    "AdviceMarksResponse",
    "AdviceReportOut",
    "AdviceResponse",
    "BacktestRunOut",
    "BacktestRunRequest",
    "BacktestRunsResponse",
    "DragonPoolItemOut",
    "DragonPoolResponse",
]

_ORM = ConfigDict(from_attributes=True)


class DragonPoolItemOut(BaseModel):
    """盘后建池候选行（``dragon_pool``，供次日开盘判定参考）。"""

    model_config = _ORM

    trade_date: date
    strategy_id: str
    code: str
    name: str | None = None
    d_date: date
    boards: int
    d_amp_pct: float | None = None
    shape_label: str | None = None
    ran_at: datetime


class DragonPoolResponse(BaseModel):
    """某交易日盘后建池候选列表。"""

    trade_date: date | None = None
    strategy_id: str | None = None
    items: list[DragonPoolItemOut] = Field(default_factory=list)


class AdviceReportOut(BaseModel):
    """策略建议报告（``payload`` 为结构化建议记录，含门槛/加分项/仓位/止损）。"""

    model_config = _ORM

    trade_date: date
    kind: str
    strategy_id: str
    strategy_version: int | None = None
    payload: dict[str, Any] = Field(default_factory=dict)
    ran_at: datetime
    created_at: datetime | None = None


class AdviceResponse(BaseModel):
    """某交易日建议报告列表。"""

    trade_date: date | None = None
    kind: str | None = None
    strategy_id: str | None = None
    items: list[AdviceReportOut] = Field(default_factory=list)


class AdviceLatestResponse(BaseModel):
    """某交易日最近一次运行的建议报告。"""

    trade_date: date | None = None
    item: AdviceReportOut | None = None


class AdviceMarkOut(BaseModel):
    """建议「已买入」标记行。"""

    model_config = _ORM

    trade_date: date
    strategy_id: str
    code: str
    marked_by: str | None = None
    marked_at: datetime | None = None


class AdviceMarksResponse(BaseModel):
    """某交易日已买入标记列表。"""

    trade_date: date
    items: list[AdviceMarkOut] = Field(default_factory=list)


class AdviceMarkRequest(BaseModel):
    """标记/取消「已买入」（取消 = 删行）。"""

    trade_date: date = Field(description="建议交易日")
    strategy_id: str = Field(min_length=1, max_length=64, description="策略标识")
    code: str = Field(min_length=1, max_length=16, description="股票代码")
    bought: bool = Field(description="true=标记已买入；false=取消")


class BacktestRunOut(BaseModel):
    """回测任务（``report`` 含 A/B/C 三段，未完成时为 ``None``）。"""

    model_config = _ORM

    run_id: str
    status: str
    start_date: date
    end_date: date
    strategies: list[str] = Field(default_factory=list)
    params: dict[str, Any] = Field(default_factory=dict)
    report: dict[str, Any] | None = None
    error: str | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    created_by: str | None = None


class BacktestRunsResponse(BaseModel):
    """最近回测任务列表（条数受 ``limit`` 约束）。"""

    limit: int
    items: list[BacktestRunOut] = Field(default_factory=list)


class BacktestRunRequest(BaseModel):
    """触发一次回测（异步执行：立即返回 running 任务行，完成/失败由后台任务更新）。"""

    start: date = Field(description="回测起始日（含）")
    end: date = Field(description="回测结束日（含）")
    strategy_id: str = Field(default="dragon", max_length=64, description="策略标识")
    params: dict[str, Any] = Field(default_factory=dict, description="策略参数覆盖")

File: frontend/src/types/report.ts
/**
 * 报告域契约：策略建议（advice_reports）与回测任务。
 *
 * 逐字段镜像后端 `app/schemas/report.py`；`payload` 为策略产出的结构化建议
 * （龙回头两路：路次 / 硬门槛 / 加分项 / 仓位 / 止损 / 卖出时点 / 字段快照）。
 */

import type { MarketMeta } from './market';

/** 建议报告行（`payload` 即结构化建议）。 */
export interface AdviceReportOut {
  trade_date: string;
  kind: string;
  strategy_id: string;
  strategy_version: number | null;
  payload: Record<string, unknown>;
  ran_at: string;
  created_at: string | null;
}

/** 某交易日建议报告列表。 */
export interface AdviceResponse extends MarketMeta {
  trade_date: string | null;
  kind: string | null;
  strategy_id: string | null;
  items: AdviceReportOut[];
}

/** 某交易日最近一次运行的建议报告。 */
export interface AdviceLatestResponse extends MarketMeta {
  trade_date: string | null;
  item: AdviceReportOut | null;
}

/** 盘后建池候选行（`dragon_pool`，Phase.POOL 识别、次日开盘判定参考）。 */
export interface DragonPoolItemOut {
  trade_date: string;
  strategy_id: string;
  code: string;
  name: string | null;
  d_date: string;
  boards: number;
  d_amp_pct: number | null;
  shape_label: string | null;
  ran_at: string;
}

/** 某交易日盘后建池候选列表。 */
export interface DragonPoolResponse extends MarketMeta {
  trade_date: string | null;
  strategy_id: string | null;
  items: DragonPoolItemOut[];
}

/** 回测任务（`report` 含 A/B/C 三段汇总，未完成为 null）。 */
export interface BacktestRunOut {
  run_id: string;
  status: string;
  start_date: string;
  end_date: string;
  strategies: string[];
  params: Record<string, unknown>;
  report: Record<string, unknown> | null;
  error: string | null;
  started_at: string | null;
  finished_at: string | null;
  created_by: string | null;
}

/** 最近回测任务列表。 */
export interface BacktestRunsResponse {
  limit: number;
  items: BacktestRunOut[];
}

/** 触发回测请求体。 */
export interface BacktestRunRequest {
  start: string;
  end: string;
  strategy_id?: string;
  params?: Record<string, unknown>;
}

// --------------------------------------------------------------- 建议载荷

/** 硬门槛命中明细。 */
export interface AdviceGate {
  factor_id: string;
  label: string;
  passed: boolean;
  detail: string | null;
}

/** 加分项命中明细。 */
export interface AdviceBonus {
  factor_id: string;
  label: string;
  satisfied: boolean;
}

/** 龙回头结构化建议（`AdviceReportOut.payload` 的具象视图）。 */
export interface DragonAdvicePayload {
  path_id: 'S2' | 'S4' | string;
  path_label: string | null;
  code: string;
  name: string | null;
  buy_day: string | null;
  buy_price: number | null;
  gates: AdviceGate[];
  bonus: AdviceBonus[];
  bonus_score: number | null;
  position: number | null;
  stop_loss_price: number | null;
  sell_timing: string | null;
  /** 复盘了结价口径：`"open"` 按 T+1 开盘价（J1 竞价抢筹）；缺省按 T+1 收盘。 */
  sell_price_ref?: 'open' | 'close';
  field_snapshot: Record<string, unknown>;
}

// --------------------------------------------------------------- 已买入标记

/** 已买入标记行（`advice_marks`）。 */
export interface AdviceMarkOut {
  trade_date: string;
  strategy_id: string;
  code: string;
  marked_by: string | null;
  marked_at: string | null;
}

/** 某交易日已买入标记列表（POST 返回全量，便于直接更新缓存）。 */
export interface AdviceMarksResponse {
  trade_date: string;
  items: AdviceMarkOut[];
}

/** 设置/取消已买入请求（取消即删行）。 */
export interface AdviceMarkRequest {
  trade_date: string;
  strategy_id: string;
  code: string;
  bought: boolean;
}
